### ISSUE TITLE
roles w/ allow_duplicates=no (default) still allows duplicates 

### ISSUE THREAD

**Issue by dghubble** [ansible/ansible#5971]
If a playbook is written to include a dependency role explicitly, that role can be loaded more than once if another role depends on it. Consider this example. `base` is a role with no dependencies. `feature` is a role with a meta/main.yml file specifying the `base` role as a dependency. `base` will be run twice by default here.

```
- name: name
  roles:
    - base
    - feature
```

The docs say "by default, roles can also only be added as a dependency once" and technically that is true, `base` is loaded once as an explicit module and once as a dependency module. Still, I argue that this behavior should be improved to "by default, roles are added only once" to be more consistent with typical dependency management systems and associated expectations. When `feature` is run, it should be known that the `base` dependency has already run.

**Comment by ansibot** [comment]
Thanks for filing a ticket! I am the friendly GitHub Ansibot. I see you did not fill out your issue description based on our new issue template. Please copy the contents of this [template](https://raw2.github.com/ansible/ansible/devel/examples/issues/ISSUE_TEMPLATE.md) and **paste it into the description** of your ticket. 

Our system will automatically close tickets that do not have an issue template filled out within 7 days. Please note that due to large interest in Ansible, humans may not comment on your ticket if you ask them questions. Don't worry, you're still in the queue and the robots are looking after you.

**Comment by mpdehaan** [comment]
Correct, right now it applies to the dependency chain.

The discussion we've had is that duplicates should apply to the same role when used with the same set of role parameters (if any) so you could still call base with different parameters.

It's a little tricky though.   It's "feature" that is choosing to allow duplicates, not the invocation of base, most likely.

So it may not always be a clear behavior.

Perhaps we'd have to introduce an "allow_duplicates: strict" or something and further explain this in the metadata comments.

cc @jimi-c

**Comment by jimi-c** [comment]
I cannot replicate this, given the following:

<pre>
# site.yml
- hosts: localhost
  connection: local
  roles:
    - base
    - feature
# roles/feature/meta/main.yml
---
allow_duplicates: false
dependencies:
  - base
# roles/feature/tasks/main.yml
---
- name: in feature
  debug: msg="here we are in feature"
# roles/base/meta/main.yml
---
allow_duplicates: false
# roles/base/tasks/main.yml
---
- name: in base
  debug: msg="here we are in base"
</pre>

The output from the above is (using the current head of devel):

<pre>
$ ansible-playbook site.yml 
PLAY [name] ******************************************************************* 
GATHERING FACTS *************************************************************** 
ok: [127.0.0.1]
TASK: [base | in base] ******************************************************** 
ok: [127.0.0.1] => {
    "item": "", 
    "msg": "here we are in base"
}
TASK: [feature | in feature] ************************************************** 
ok: [127.0.0.1] => {
    "item": "", 
    "msg": "here we are in feature"
}
PLAY RECAP ******************************************************************** 
127.0.0.1                  : ok=3    changed=0    unreachable=0    failed=0   
</pre>

As far as I'm aware, the only issue with de-duplication of roles involves using roles across multiple plays, since the deps are tracked per-play only.

Could you provide an example that replicates the issue?

**Comment by jimi-c** [comment]
Fixed a typo above (`s/allow_dependencies/allow_duplicates/g`), but since the default is to not allow dupes, the correct output was produced anyway.
